table/tables/tables.go

```diff
@@ -407,18 +448,19 @@ func (t *Table) getRollbackableMemStore(ctx sessionctx.Context) kv.RetrieverMuta
 // locatePartition returns the partition ID of the input record.
 func (t *Table) locatePartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int64, error) {
 	// TODO: Remove this later, when we have better way to TestPartitionAddRecord.
-	if t.partitionExpr == nil {
-		partitionExpr, err := generatePartitionExpr(t.meta)
+	if t.partitionExprCache == nil {
```

Comment:
Is there concurrency issue?

Reply:
Read and write will not happen at the same time.
This line will just be used in test, notice the TODO.